src/Broadway/EventStore/DBALEventStore.php

```diff
@@ -231,4 +235,80 @@ private function convertStorageValueToIdentifier($id)
 
         return $id;
     }
+
+    public function visitEvents(EventVisitorInterface $eventVisitor, Criteria $criteria = null)
+    {
+        $statement = $this->prepareVisitEventsStatement($criteria ?: Criteria::create());
+        $statement->execute();
+
+        while ($row = $statement->fetch()) {
+            if ($this->useBinary) {
+                $row['uuid'] = $this->convertStorageValueToIdentifier($row['uuid']);
+            }
+
+            $domainMessage = $this->deserializeEvent($row);
+
+            $eventVisitor->doWithEvent($domainMessage);
+        }
+    }
+
+    private function prepareVisitEventsStatement(Criteria $criteria)
+    {
+        list ($where, $bindValues, $bindValueTypes) = $this->prepareVisitEventsStatementWhereAndBindValues($criteria);
+        $query = 'SELECT uuid, playhead, metadata, payload, recorded_on
+            FROM ' . $this->tableName . '
+            ' . $where . '
+            ORDER BY id ASC';
```

Comment:
Can you move the logic to go from `Criteria` to its own class? Something like:
- `DBALCriteriaQuery($query, $bindValues, $bindValuesTypes);`
- `DBALCriteriaQueryBuilder`

I guess instead of having a builder class it could also be a static method on the `DBALCriteriaQuery`.